### PR TITLE
Remove SingletonContextFactory database testing interface

### PR DESCRIPTION
**Pull request by peppy**
It is dangerous to use this as it doesn't correctly handle contexts and can cause issues that will never actually arise in normal execution.

While working on tests I came across what seemed to be a bug in the database subsystem, but turned out to be that every context was being incorrectly shared.

---

### CODE REVIEW

File: osu.Game/Tests/Platform/TestStorage.cs [hunk 1]
```diff
@@ -7,13 +7,11 @@ namespace osu.Game.Tests.Platform
...
+        public TestStorage(string baseName)
+            : base(baseName, null)
         {
         }
 
-        public override string GetDatabaseConnectionString(string name)
-        {
-            return "DataSource=:memory:";
-        }
+        public override string GetDatabaseConnectionString(string name) => "Data Source=" + GetUsablePathFor($"{(object)name}.db", true);
```
Comment (smoogipoo) [inline osu.Game/Tests/Platform/TestStorage.cs]: `(object)name` ?

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game/Tests/Platform/TestStorage.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Tests/Platform/TestStorage.cs]: weird, that's how it is in the other class. probably left over from some old usage pattern

File: osu.Game/Tests/Platform/TestStorage.cs [hunk 2]
```diff
@@ -7,13 +7,11 @@ namespace osu.Game.Tests.Platform
...
+        public TestStorage(string baseName)
+            : base(baseName, null)
         {
         }
 
-        public override string GetDatabaseConnectionString(string name)
-        {
-            return "DataSource=:memory:";
-        }
+        public override string GetDatabaseConnectionString(string name) => "Data Source=" + GetUsablePathFor($"{name}.db", true);
```
Comment (smoogipoo) [inline osu.Game/Tests/Platform/TestStorage.cs]: At this point "TestStorage" serves no purpose anymore. This looks to be the same thing that `DesktopStorage` does.

Is there a reason you kept it?

File: osu.Game.Tests/Visual/TestCasePlaySongSelect.cs [hunk 3]
```diff
@@ -62,7 +62,12 @@ private void load()
             var storage = new TestStorage(@"TestCasePlaySongSelect");
```
Comment (smoogipoo) [inline osu.Game.Tests/Visual/TestCasePlaySongSelect.cs]: This should probably be guid'd now, clicking on `TestCasePlaySongSelect` multiple times can cause crashes:

`Microsoft.Data.Sqlite.SqliteException: SQLite Error 10: 'disk I/O error'.`

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Tests/Platform/TestStorage.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Tests/Platform/TestStorage.cs]: good point

Comment (Aergwyn) [reply osu.Game/Tests/Platform/TestStorage.cs]: Was it intended to go away from an in-memory database?

Comment (peppy) [reply osu.Game/Tests/Platform/TestStorage.cs]: yes, that's the point of this branch (in-memory cannot handle multiple contexts; it creates a fresh database per context).

**Review by smoogipoo — APPROVED** [review]
Seems good to me.

**Pull request merged**